File: cs/algorithms_in/recursive_binary_search.cs
using System;

namespace algorithms_in;

public class RecursiveBinarySearch
{
        public static bool recursiveBinarySearch(int[] array, int l, int r, int query) {
            if (l > r) return false;

            int i = l + (r - l) / 2;
            if (array[i] == query) return true;
            if (array[i] < query) {
                return recursiveBinarySearch(array, i + 1, r, query);
            }
            if (array[i] > query) {
                return recursiveBinarySearch(array, l, i - 1, query);
            }

            return false;
        }
}

File: rust/algorithms_in/src/recursive_binary_search.rs
pub fn recursive_binary_search(slice: &[usize], l: usize, r: usize, query: usize) -> bool {
    if l > r { return false };

    let i: usize = l + (r - l) / 2;

    if slice[i] == query { return true };
    if slice[i] < query {
        return recursive_binary_search(slice, i + 1, r, query);
    }
    if slice[i] > query {
        return recursive_binary_search(slice, l, i - 1, query);
    }

    return false;
}